### PR TITLE
Rate provider pallet

### PR DESCRIPTION
**Pull request by stanly-johnson**
This pallet allows Liquidity providers to publish rates on chain for off-chain/on-chain transactions.

related #87

### CODE REVIEW

File: pallets/rate_provider/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn rates)]
+    // Rates provided by an LP, the actual rate is not stored here
+    // this only represents the basis points above/below the rates supplied by oracle
+    // module
+    pub(super) type RateStore<T: Config> = StorageDoubleMap<
```
[the remark this answers is not in the document]
Comment (olanod) [reply pallets/rate_provider/src/lib.rs:65]: To elaborate more on the types of the example, is often more common to use this base/quote naming than form/to and more confusing to mix(form->base?) so we could have a
```rust
struct AssetPair<B, Q> {
    pub base: B,
    pub quote: Q,
}
```
And as convention we would use for example the crypto always as base in our runtime configuration and everywhere so is easier to remember what is what.

For the map I also think we can put our percentage inside a `RateInfo`/`RateDetail`(or just "`Rate`") struct, so later we can add more features like allowing providers to specify a min/max amount they can transact with, fixed costs, maybe the future option of also setting rate by exact price and even encode a "price function" that gives us a rate as function of an amount. Also perhaps a good place to add your combine rates functionality by implementing the `core::ops::Add` trait? Rate might even be an enum so you could do `Rate::Fixed(123) + Rate::Premium(1)` 🤔 wdyt?

File: pallets/rate_provider/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    use orml_traits::DataProvider;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        /// Type of assets that the LP can set rates for
+        type CurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord;
+        /// Type of the base currency
+        type BaseCurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord;
```
Comment (olanod) [inline pallets/rate_provider/src/lib.rs:46]: `Asset` and `BaseAsset`? 🙂

File: pallets/rate_provider/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        type Whitelist: Contains<Self::AccountId>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn rates)]
+    // Rates provided by an LP, the actual rate is not stored here
```
Comment (olanod) [inline pallets/rate_provider/src/lib.rs:62]: In the context of this pallet let's just call them "providers" as the liquidity part is more of a Valiu concept that might not apply for other projects

**Review by olanod — CHANGES_REQUESTED** [review]

File: pallets/rate_provider/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        /// Type of assets that the LP can set rates for
+        type CurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord;
+        /// Type of the base currency
+        type BaseCurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord;
+        /// type of Oracle Provider
+        type OracleProvider: DataProvider<Self::CurrencyId, Self::OracleValue>;
```
[the remark this answers is not in the document]
Comment (stanly-johnson) [reply pallets/rate_provider/src/lib.rs:48]: Are you talking about renaming `OracleProvider` -> `PriceFeed` or replacing `DataProvider` with [DataFeeder](https://github.com/open-web3-stack/open-runtime-module-library/blob/b63b6667a0809165bda20ba57cbe85e1c0687b8c/oracle/src/lib.rs#L323) 🤔

[the remark this answers is not in the document]
Comment (olanod) [reply pallets/rate_provider/src/lib.rs]: Renaming the OracleProvider, the DataProvider bound is the way to go :)

File: pallets/rate_provider/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (stanly-johnson) [reply pallets/rate_provider/src/lib.rs]: > querying for example the list of mediums for a given pair or the list of providers and their rates for a pair+medium which is what I suspect we will be doing more often

I've updated the storage to BTreeMap and added `get_pair_rates` to `RateProvider` to  fetch the list of providers/rates.

[the remark this answers is not in the document]
Comment (stanly-johnson) [reply pallets/rate_provider/src/lib.rs]: > For the map I also think we can put our percentage inside a RateInfo/RateDetail(or just "Rate") struct, so later we can add more features like allowing providers to specify a min/max amount they can transact with, fixed costs, maybe the future option of also setting rate by exact price and even encode a "price function" that gives us a rate as function of an amount. Also perhaps a good place to add your combine rates functionality by implementing the core::ops::Add trait

I've converted it to a struct `RateDetail`, and added `RateCombinator` type in the pallet to get access to combine_rate function. I kinda feel this is much more easier for someone who is implementing this pallet for their runtime, also this is similar to the implementation of `orml_oracle`

**Review by olanod — CHANGES_REQUESTED** [review]
Looking good ;) just a few remarks

File: primitives/src/rates.rs [hunk 5]
```diff
@@ -11,18 +35,21 @@ pub enum PaymentMethod {
 }
 
 // A trait for querying rates supplied by an LP
-pub trait RateProvider<F, T, M, Z, R, P> {
-    fn get_rates(from: F, to: T, method: M, who: Z) -> Option<(R, P)>;
+pub trait RateProvider<X, M, Z, R, T> {
```
[the remark this answers is not in the document]
Comment (stanly-johnson) [reply primitives/src/rates.rs]: > This trait might be tooo generic

😆 Fixed

File: pallets/rate_provider/src/lib.rs [hunk 6]
```diff
@@ -35,7 +36,7 @@ pub mod pallet {
         /// type of Oracle Provider
         type PriceFeed: DataProvider<Self::Asset, Self::OracleValue>;
         /// type of Oracle Value
-        type OracleValue: Parameter + Member + Ord;
+        type OracleValue: Parameter + Member + Ord + FixedPointNumber;
```
Comment (olanod) [inline pallets/rate_provider/src/lib.rs:39]:
```suggestion
        type OracleValue: Parameter + Member + FixedPointNumber;
```

**Review by olanod — APPROVED** [review]

**Pull request merged**